### PR TITLE
Chapter 2: Dapr @ 20000 feet

### PR DESCRIPTION
**Pull request by amolenk**
Hi Robert,

We created a new release PR for chapter 2 because the original `dapr/release/ch2` branch already contained most of the chapter 2 content, which makes reviewing rather cumbersome because we can only comment on added lines through the UI.

### CODE REVIEW

File: docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+![Dapr traffic patterns](./media/dapr-traffic-patterns.png)
+
+**Figure 2-7**. Dapr traffic patterns
+
+Looking at the previous figure, one might question the latency and overhead incurred for each call.  
+
+The Dapr team has invested heavily in performance. A tremendous amount of engineering effort has gone into to making Dapr efficient. Calls between Dapr sidecars are always made with gRPC, which delivers high performance and small, binary payloads. In most cases, the additional overhead should be less than 1 millisecond. 
+
+To increase performance, developers can implement calls to Dapr building blocks with gRPC.
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md:246]:
```suggestion
To increase performance, developers can call the Dapr building blocks with gRPC. 
```

File: docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+- Multiplexing support for sending multiple parallel requests over the same connection - HTTP 1.1 limits processing to one request/response message at a time.
+- Bidirectional full-duplex communication for sending both client requests and server responses simultaneously.
+- Built-in streaming enabling requests and responses to asynchronously stream large data sets.
+
+## Dapr and service meshes
+
+Service mesh is another rapidly evolving technology for distributed applications. 
+
+A service mesh is a configurable infrastructure layer with built-in capabilities to handle service-to-service communication, resiliency, load balancing, and telemetry capture. It moves the responsibility for these concerns out of the microservices and into service mesh layer. Like Dapr, a service mesh also exposes a sidecar architecture.
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md:258]:
```suggestion
A service mesh is a configurable infrastructure layer with built-in capabilities to handle service-to-service communication, resiliency, load balancing, and telemetry capture. It moves the responsibility for these concerns out of the microservices and into the service mesh layer. Like Dapr, a service mesh also follows a sidecar architecture.
```

File: docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+Figure 2-8 shows an application that implements service mesh technology.
+
+![Service Mesh](./media/service-mesh-with-side-car.png)
+
+**Figure 2-8**. Service mesh with a side car.
+
+Note in the previous figure how messages are intercepted by a proxy that runs alongside each microservice. Each proxy can be configured with traffic rules specific to the microservice. It understands messages and can route them across your services and the outside world.
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md:266]: *Service* and *microservice* are both used in the text. Should we decide on one throughout the text/book? If so, which do you prefer?

File: docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+
+While both technologies use a sidecar architecture, each is different. Dapr provides distributed application features. A service mesh provides a dedicated network infrastructure layer.
+
+As each technology works at a different level, both can work together in the same application. Connected together, application traffic would route to the Dapr sidecar first and then the service mesh sidecar and then out. On the other end, traffic would traverse the stack in reverse order: Service mesh to Dapr to application. 
+
+Figure 2-9 shows an application that implements both Dapr and service mesh technology.
+
+![Dapr and Service Mesh together](./media/dapr-and-service-mesh.png)
+
+**Figure 2-9**. Dapr and service mesh together.
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md:278]: While it is explained in the Haishi/Yaron book this way, we don't think service invocation is the best way to show how Dapr and a Service Mesh can work together. For this to work, you would probably need to write your own Dapr service discovery component to let the Dapr sidecar route the request to the service mesh sidecar. Even then, you won't gain that much value as Linkerd already does service-to-service invocation pretty well.
We can show the value better by making an example where Service 1 has two sidecars. For example, Service 1 can use a Linkerd sidecar to make calls to Service 2, while at the same time use a Dapr sidecar to store contextual state (or do pub/sub, or use actors).
This way, the sidecars aren't connected together, but both serve the service independently.

**Review by amolenk — COMMENTED** [review]
Hi Robert,
Edwin and I reviewed this together and went into hypercritical mode to ensure we got the fundamental terminology just right. In a way I think this is a much more difficult chapter to write than the building block chapters because it's so important that the reader gets a good grip on the different concepts and how they relate. We've also put some questions in the comments to discuss in our next meeting.

File: docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md (hunk 2 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md]: Added

File: docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md (hunk 3 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/dapr-at-20000-feet.md]: Let's go with the rule from our discussion. Microservice in the context of an application. Service when speaking of single service.

**Pull request closed without merging**